### PR TITLE
Fqe prune columns

### PR DESCRIPTION
**Pull request by ZoranPandovski**
## Description

MindsDB would fetch all columns from source databases (using SELECT *), even when the query only needed a few columns. 
Now,  it analyzes the query and only fetches the columns we  actually need.
```
--e.g
SELECT customer_name, order_total 
FROM snowflake.customers 
JOIN bigquery.orders ON customers.id = orders.customer_id

-- MindsDB fetched:
SELECT * FROM customers  -- All 50 columns! 
SELECT * FROM orders      -- All 30 columns!

--Now:
SELECT id, customer_name FROM customers  
SELECT customer_id, order_total FROM orders 
```
* Before: Full table scans fetched all columns (e.g 20 + 15 columns 90 MB total x 100,000 rows), leading to ~8.5 s network time and ~1.2 GB memory usage.
* After: Only required columns are fetched (e.g 4 + 3 18 MB total), reducing network time to ~1.8 s and memory usage to ~180 MB.
* Result: ~80% reduction in data size, faster execution, lower memory footprint, and improved overall efficiency.

This optimization is especially impactful for wide tables and large joins.

fixes https://linear.app/mindsdb/issue/FQE-1741/column-pruning

## Type of change

(Please delete options that are not relevant)

- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [X] ⚡ New feature (non-breaking change which adds functionality)
- [ ] 📢 Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] 📄 This change requires a documentation update

## Checklist:

- [X] My code follows the style guidelines(PEP 8) of MindsDB.
- [X] I have appropriately commented on my code, especially in complex areas.
- [ ] Necessary documentation updates are either made or tracked in issues.
- [X] Relevant unit and integration tests are updated or added.

### CODE REVIEW

File: tests/unit/planner/test_column_pruning.py [hunk 1]
```diff
@@ -0,0 +1,388 @@
+from mindsdb_sql_parser import parse_sql
+from mindsdb_sql_parser.ast import Identifier, Select
+
+from mindsdb.api.executor.planner import plan_query
+
+
+class TestColumnPruning:
+    """Test column pruning optimization."""
+    
+    def test_basic_column_pruning(self):
```
[the remark this answers is not in the document]
Comment (ZoranPandovski) [reply tests/unit/planner/test_column_pruning.py:11]: Wrong

File: tests/unit/planner/test_injected_data.py [hunk 2]
```diff
@@ -1,100 +1,88 @@
 import copy
 
-from mindsdb_sql_parser.ast import (Identifier, Select, Join, Constant, Star, Data, BinaryOperation)
+from mindsdb_sql_parser.ast import Identifier, Select, Join, Constant, Data, BinaryOperation
 from mindsdb_sql_parser.utils import JoinType
 
 from mindsdb.api.executor.planner import plan_query
```
[the remark this answers is not in the document]
Comment (ZoranPandovski) [reply tests/unit/planner/test_injected_data.py:6]: Wrong

File: tests/unit/planner/test_join_tables.py [hunk 3]
```diff
@@ -3,7 +3,16 @@
 import pytest
 
 from mindsdb_sql_parser import parse_sql
-from mindsdb_sql_parser.ast import Identifier, Select, Join, Constant, Star, BinaryOperation, OrderBy, Function
+from mindsdb_sql_parser.ast import (
+    Identifier,
+    Select,
+    Join,
```
[the remark this answers is not in the document]
Comment (ZoranPandovski) [reply tests/unit/planner/test_join_tables.py:9]: Wrong

File: tests/unit/planner/test_plan_union.py [hunk 4]
```diff
@@ -1,87 +1,81 @@
 from mindsdb_sql_parser import parse_sql
-from mindsdb_sql_parser.ast import (Identifier, Select, Join, Constant, Star, Union, BinaryOperation)
+from mindsdb_sql_parser.ast import Identifier, Select, Join, Constant, Union, BinaryOperation
 from mindsdb_sql_parser.utils import JoinType
 
 from mindsdb.api.executor.planner import plan_query
 from mindsdb.api.executor.planner.query_plan import QueryPlan
```
[the remark this answers is not in the document]
Comment (ZoranPandovski) [reply tests/unit/planner/test_plan_union.py:6]: Wrong

File: mindsdb/api/executor/planner/plan_join.py [hunk 5]
```diff
@@ -431,15 +410,151 @@ def process_subselect(self, item):
...
+                    result[0] = True
+
+            query_traversal(query_in.where, check_last)
+            if result[0]:
+                return None
+
+        columns = {}
+        has_qualified_star_for_table = False
+
+        alias_map = {}
```
[the remark this answers is not in the document]
Comment (ZoranPandovski) [reply mindsdb/api/executor/planner/plan_join.py]: Wrong, it is imported line 16

File: mindsdb/api/executor/planner/plan_join.py [hunk 6]
```diff
@@ -431,15 +410,157 @@ def process_subselect(self, item):
...
+        Returns a list of column Identifiers or None if we should fetch all columns.
+        """
+        # Skip column pruning for predictors/models
+        if table_info.predictor_info is not None:
+            return None
+        
+        # Skip column pruning for subselects with custom projections
+        # These already have their SELECT clause defined
+        if table_info.sub_select is not None and isinstance(table_info.sub_select, Select):
+            if table_info.sub_select.targets and not any(isinstance(t, Star) for t in table_info.sub_select.targets):
```
[the remark this answers is not in the document]
Comment (ZoranPandovski) [reply mindsdb/api/executor/planner/plan_join.py]: The truthiness check table_info.sub_select.targets already handles None and empty cases correctly

[the remark this answers is not in the document]
Comment (ZoranPandovski) [reply mindsdb/api/executor/planner/plan_join.py]: any() uses short-circuit evaluation and is optimal for this use case

File: mindsdb/api/executor/planner/plan_join.py [hunk 7]
```diff
@@ -431,15 +410,157 @@ def process_subselect(self, item):
...
+                return None
```
[the remark this answers is not in the document]
Comment (ZoranPandovski) [reply mindsdb/api/executor/planner/plan_join.py]: Star is properly imported and isinstance() works correctly

File: mindsdb/api/executor/planner/plan_join.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ZoranPandovski) [reply mindsdb/api/executor/planner/plan_join.py]: The edge case (SELECT *, col1) is real but the code handles it correctly

[the remark this answers is not in the document]
Comment (ZoranPandovski) [reply mindsdb/api/executor/planner/plan_join.py]: I will add test for this

File: mindsdb/api/executor/planner/plan_join.py [hunk 8]
```diff
@@ -279,49 +278,24 @@ def _check_node_condition(node, **kwargs):
...
+                        regular_table_count += 1
                 elif isinstance(item, Join):
-                    join_type = (
-                        item.join_type.upper() if hasattr(item.join_type, "upper") else str(item.join_type).upper()
-                    )
-
-                    # LEFT JOIN preserves left table row count - LIMIT pushdown is safe
-                    if join_type in ("LEFT JOIN", "LEFT OUTER JOIN"):
-                        continue
+                    has_join = True
```
Comment (ea-rus) [inline mindsdb/api/executor/planner/plan_join.py]: `if join_type in ("LEFT JOIN", "LEFT OUTER JOIN"):` - this condition was removed
is it braking something?

File: mindsdb/api/executor/planner/plan_join.py [hunk 9]
```diff
@@ -431,15 +416,194 @@ def process_subselect(self, item):
...
+        Collect all columns needed from a specific table for column pruning optimization.
+
+        Note: Caller should check _can_prune_columns() before calling this method.
+
+        Returns a list of column Identifiers or None if we should fetch all columns.
+        """
+        # Skip column pruning if query uses LAST
+        # TODO: We need to improve this
+        if query_in.where:
+            result = [False]
```
Comment (ea-rus) [inline mindsdb/api/executor/planner/plan_join.py]: this block isn't used
I removed it and tests are passed

File: mindsdb/utilities/render/sqlalchemy_render.py [hunk 10]
```diff
@@ -213,7 +213,13 @@ def to_expression(self, t):
             if col is None:
                 col = self.to_column(t)
             if t.alias:
-                col = col.label(self.get_alias(t.alias))
+                alias_name = self.get_alias(t.alias)
+                # Skip self-referencing aliases (e.g., "column AS column")
+                # Some databases like DuckDB don't support them
```
Comment (ea-rus) [inline mindsdb/utilities/render/sqlalchemy_render.py]: duckdb should suport it, isn't?

File: tests/unit/handlers/test_mysql.py [hunk 11]
```diff
@@ -220,6 +220,80 @@ def test_disconnect(self):
         self.handler.disconnect()
         mock_conn.close.assert_not_called()
 
+    def test_check_connection_success(self):
```
Comment (ea-rus) [inline tests/unit/handlers/test_mysql.py:223]: is mysql handler update related to this PR?

File: tests/unit/planner/test_column_pruning.py [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+        """)
+
+        plan = plan_query(query, integrations=["int1", "int2"])
+
+        # Subselects with pure SELECT * should be pruned to only needed columns
+        found_pruned_subselect = False
+        for step in plan.steps:
+            step_str = str(step)
+            # Look for SubSelectStep with id column (not SELECT *)
+            if "SubSelect" in step_str and ("id" in step_str) and "SELECT *" not in step_str:
```
Comment (ea-rus) [inline tests/unit/planner/test_column_pruning.py]: is better to check type: isinstance(step, SubSelectStep)

Comment (ZoranPandovski) [reply tests/unit/planner/test_column_pruning.py]: Good catch, changing this now

File: tests/unit/handlers/test_mysql.py (hunk 11 above)
Comment (ZoranPandovski) [reply tests/unit/handlers/test_mysql.py:223]: Not sure what happend but I can see:
```
[Merge branch 'FQE-1383' into fqe-prune-columns](https://github.com/mindsdb/mindsdb/pull/11775/commits/d5218a0623de4429cfbb683a8aac35d2ff3410b6)
```

File: mindsdb/api/executor/planner/plan_join.py (hunk 9 above)
Comment (ZoranPandovski) [reply mindsdb/api/executor/planner/plan_join.py]: Let me double check this since some of the scenarios with LAST were failing before

Comment (ZoranPandovski) [reply mindsdb/api/executor/planner/plan_join.py]: I checked the scenario but it was more related to the predictor then LAST, all test pass now so I will remove this safeguard

File: mindsdb/utilities/render/sqlalchemy_render.py (hunk 10 above)
Comment (ZoranPandovski) [reply mindsdb/utilities/render/sqlalchemy_render.py]: I think we may remove the comment to make it more clear. The issue was that with some tests I think KB we got:
duckdb.duckdb.BinderException: Binder Error: Column "content" referenced that ... So we just  check if alias name equals column name. If yes, skip

File: mindsdb/api/executor/planner/plan_join.py (hunk 8 above)
Comment (ZoranPandovski) [reply mindsdb/api/executor/planner/plan_join.py]: Its not breaking since it is safe due to LEFT joins is safe for LIMIT but may be slower. I think we can add back this optimization

File: tests/unit/handlers/test_mysql.py (hunk 11 above)
Comment (ZoranPandovski) [reply tests/unit/handlers/test_mysql.py:223]: I think Lucas  revert this

**Review by hamishfagg — APPROVED** [review]
